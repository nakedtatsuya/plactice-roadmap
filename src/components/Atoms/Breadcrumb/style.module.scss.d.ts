declare const styles: {
  readonly "root": string;
  readonly "ol": string;
  readonly "separator": string;
  readonly "ellipsis": string;
};
export = styles;

