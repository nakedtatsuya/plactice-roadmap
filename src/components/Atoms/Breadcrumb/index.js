import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.scss'
import {containPresenter} from '../../utils/HoC'

const BreadcrumbPresenter = ({tag: Tag, children, ...props}) => (
  <Tag {...props}>
    <ol className={styles.ol}>{children}</ol>
  </Tag>
)

BreadcrumbPresenter.propTypes = {
  tag: PropTypes.string,
  children: PropTypes.node.isRequired,
}

/**
 * コンテナー
 * childrenの要素数に応じてリストを生成してpresenterに渡す
 */
export const BreadcrumbContainer = ({
  children,
  tag,
  separator,
  presenter,
  ...props
}) => {
  const items = React.Children.toArray(children)
    .filter(child => React.isValidElement(child))
    .map((child, index) => (
      <li className={styles.li} key={`child-${index}`}>
        {child}
      </li>
    ))

  return presenter({
    tag,
    children: insertSeparators(items, separator),
    ...props,
  })
}

const Breadcrumb = containPresenter(BreadcrumbContainer, BreadcrumbPresenter)

Breadcrumb.defaultProps = {
  separator: '/',
  tag: 'nav',
}

export default Breadcrumb

/**
 * ヘルパー関数
 * メニューの間にセパレイト文字をいれる
 * @param items
 * @param separator
 */
export const insertSeparators = (items, separator) => {
  return items.reduce((acc, current, index) => {
    if (index < items.length - 1) {
      acc = acc.concat(
        current,
        <li
          aria-hidden={true}
          key={`separator-${index}`}
          className={styles.separator}
        >
          {separator}
        </li>
      )
    } else {
      acc.push(current)
    }

    return acc
  }, [])
}
