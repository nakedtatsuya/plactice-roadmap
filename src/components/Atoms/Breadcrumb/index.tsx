import React, { useState, ReactNodeArray } from 'react';
import style from './style.module.scss';
import { containPresenter } from '../../utils/HoC';

interface IProps {
  maxItems?: number;
  itemsAfterCollapse?: number;
  itemsBeforeCollapse?: number;
  separator?: React.ReactNode;
  className?: string;
  tag?: 'div' | 'nav';
}

interface PresenterProps {
  className: string;
  children: React.ReactNode;
  tag: 'div' | 'nav';
}

interface ContainerProps extends IProps {
  presenter: (presenter: PresenterProps) => JSX.Element;
}

/**
 * プレゼンター
 * コンテナーから渡されたメニューを表示する
 * @param items
 * @param separator
 */
const BreadcrumbPresenter: React.FC<PresenterProps> = ({
  tag: Tag,
  children,
  className,
}) => (
  <Tag className={[style.root, className].join(' ')}>
    <ol className={style.ol}>{children}</ol>
  </Tag>
);

/**
 * コンテナー
 * メニューに応じてリストを生成してpresenterを呼び出す
 * @param items
 * @param separator
 */
const BreadcrumbContainer: React.FC<ContainerProps> = ({
  children,
  className,
  tag,
  itemsAfterCollapse,
  itemsBeforeCollapse,
  maxItems,
  separator,
  presenter,
}) => {

  const [expanded, setExpanded] = useState(false);

  /**
   * ヘルパー関数
   * 指定された数だけ最初と最後のITEMを残してそれ以外の中間の項目を省略
   * 現状の長さで問題ないときはそのまま返す
   */
  const renderItemsBeforeAndAfter = (
    allItems: ReactNodeArray
  ): ReactNodeArray => {
    const handleClickExpand = () => {
      setExpanded(true);
    };

    if (itemsBeforeCollapse + itemsAfterCollapse >= allItems.length) {
      return allItems;
    }

    return [
      ...allItems.slice(0, itemsBeforeCollapse),
      <li key={'ellipsis'} className={style.ellipsis}><button onClick={handleClickExpand}>...</button></li>,
      ...allItems.slice(allItems.length - itemsAfterCollapse, allItems.length),
    ];
  };

  /**
   * ヘルパー関数
   * メニューの間にセパレイト文字をいれる
   * @param items
   * @param separator
   */
  const insertSeparators = (
    items: React.ReactNodeArray,
    separator: React.ReactNode
  ) => {
    return items.reduce(
      (acc: React.ReactNodeArray, current: React.ReactNode, index: number) => {
        if (index < items.length - 1) {
          acc = acc.concat(
            current,
            <li
              aria-hidden
              key={`separator-${index}`}
              className={style.separator}
            >
              {separator}
            </li>
          );
        } else {
          acc.push(current);
        }

        return acc;
      },
      []
    );
  };


  /**
   * main 処理
   */
  const allItems = React.Children.toArray(children)
    .filter(child => {
      return React.isValidElement(child);
    })
    .map((child, index) => <li key={`child-${index}`}>{child}</li>);

  return presenter({
    tag,
    children: insertSeparators(
      expanded || allItems.length <= maxItems
        ? allItems
        : renderItemsBeforeAndAfter(
            allItems
          ),
      separator
    ),
    className,
  });
};

/**
 * Breadcrumb
 * メニューの間にセパレイト文字をいれる
 * @param items
 * @param separator
 */
const Breadcrumb: React.FC<IProps> = containPresenter(
  BreadcrumbContainer,
  BreadcrumbPresenter
);

Breadcrumb.defaultProps = {
  maxItems: 8,
  itemsAfterCollapse: 1,
  itemsBeforeCollapse: 1,
  separator: '/',
  tag: 'nav',
};

export default Breadcrumb;



