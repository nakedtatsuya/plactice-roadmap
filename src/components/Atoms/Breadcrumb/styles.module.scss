.ol {
  align-items: center;
  display: flex;
  list-style: none;
  overflow: hidden;
  width: 100%;
}

.li {
  max-width: 50%;

  > p,
  > a {
    display: block;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    width: 100%;
  }
}

.separator {
  display: flex;
  margin-left: var(--space-12);
  margin-right: var(--space-12);
  user-select: none;
}
