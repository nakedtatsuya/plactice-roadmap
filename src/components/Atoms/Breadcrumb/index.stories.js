import React from 'react';
import { storiesOf } from '@storybook/react';
import Breadcrumb from './index';

storiesOf(`Atoms/Breadcrumb`, module)
.add(`default`, () => (
  <Breadcrumb>
    <a>item 1</a>
    <a>item 3</a>
    <a>item 2</a>
  </Breadcrumb>
))
.add(`１アイテム`, () => (
  <Breadcrumb>
    <a>item 1</a>
  </Breadcrumb>
))
.add(`max Item`, () => (
  <Breadcrumb maxItems={9}>
    <a>item 1</a>
    <a>item 2</a>
    <a>item 3</a>
    <a>item 4</a>
    <a>item 5</a>
    <a>item 6</a>
    <a>item 7</a>
    <a>item 8</a>
    <a>item 9</a>
  </Breadcrumb>
))
.add(`itemsAfterCollapse: 3`, () => (
  <Breadcrumb maxItems={4} itemsAfterCollapse={3}>
    <a>item 1</a>
    <a>item 2</a>
    <a>item 3</a>
    <a>item 4</a>
    <a>item 5</a>
  </Breadcrumb>
))
.add(`itemsBeforeCollapse: 3`, () => (
  <Breadcrumb itemsBeforeCollapse={3}>
    <a>item 1</a>
    <a>item 2</a>
    <a>item 3</a>
    <a>item 4</a>
    <a>item 5</a>
    <a>item 6</a>
    <a>item 7</a>
    <a>item 8</a>
    <a>item 9</a>
  </Breadcrumb>
))
.add(`custom separator`, () => (
  <Breadcrumb separator={'>'}>
    <a>item 1</a>
    <a>item 2</a>
    <a>item 3</a>
  </Breadcrumb>
));