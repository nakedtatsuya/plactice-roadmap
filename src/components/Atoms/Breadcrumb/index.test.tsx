import React from 'react';
import renderer from 'react-test-renderer';
import Breadcrumb from './index';

describe('Breadcrumb', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Breadcrumb>
          <a>list</a>
          <a>list</a>
          <a>list</a>
        </Breadcrumb>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Breadcrumb props', () => {
  it('className base', () => {
    const classArray = renderer
      .create(
        <Breadcrumb className={'addClass'}>
          <a>list</a>
          <a>list</a>
          <a>list</a>
        </Breadcrumb>
      )
      .toJSON()
      .props.className.split(' ');
    expect('root').toMatch(classArray[0]);
    expect('addClass').toMatch(classArray[1]);
  });
});
