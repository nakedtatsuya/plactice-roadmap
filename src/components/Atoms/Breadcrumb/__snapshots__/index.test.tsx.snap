// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders correctly 1`] = `
<nav
  className="root "
>
  <ol
    className="ol"
  >
    <li>
      <a>
        list
      </a>
    </li>
    <li
      aria-hidden={true}
      className="separator"
    >
      /
    </li>
    <li>
      <a>
        list
      </a>
    </li>
    <li
      aria-hidden={true}
      className="separator"
    >
      /
    </li>
    <li>
      <a>
        list
      </a>
    </li>
  </ol>
</nav>
`;
